rule download_human_fasta:
    output:
        config['genome']['fasta']['hg38']
    shell:
        "wget -O - https://www.encodeproject.org/files/GRCh38_no_alt_analysis_set_GCA_000001405.15/@@download/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta.gz | gunzip -c > {output}"


rule download_human_gtf:
    output:
        config['genome']['gtf_raw']['hg38']
    shell:
        "wget -O - http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_38/gencode.v38.annotation.gtf.gz | gunzip -c > {output}"


rule download_fasta_mm10:
    output:
        config['genome']['fasta']['mm10']
    shell:
        "wget -O - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/GRCm38.primary_assembly.genome.fa.gz | gunzip -c > {output}"


rule download_gtf_mm10:
    output:
        config['genome']['gtf_raw']['mm10']
    shell:
        "wget -O - https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/gencode.vM25.annotation.gtf.gz | gunzip -c > {output}"


rule download_encode_fastq:
    threads: 1
    resources:
        mem_gb = 4
    output:
        config['encode']['fastq']
    shell:
        "wget https://www.encodeproject.org/files/{wildcards.encode_id}/@@download/{wildcards.encode_id}.fastq.gz -O {output}"


rule download_encode_bam:
    threads: 1
    resources:
        mem_gb = 4
    output:
        bam = config['encode']['bam']
    shell:
        "wget https://www.encodeproject.org/files/{wildcards.encode_id}/@@download/{wildcards.encode_id}.bam -O {output}"


rule download_gene_exp:
    output:
        config['short_rnaseq']['gene_exp'].format(sample='WTC11')
    shell:
        "wget https://www.encodeproject.org/files/ENCFF178KKE/@@download/ENCFF178KKE.tsv -O {output}"
